Sources/FoundationEssentials/JSON/OutputBuffer.swift

```diff
@@ -0,0 +1,250 @@
+    ) {
+        guard let sourceAddress = source.baseAddress, !source.isEmpty else {
+            return
+        }
+        let available = capacity &- initialized
+        precondition(
+            source.count <= available,
+            "buffer cannot contain every element from source."
+        )
+        let tail = start.advanced(by: initialized)
+        tail.moveInitialize(from: sourceAddress, count: source.count)
+        initialized &+= source.count
+    }
+
+    mutating func moveAppend(
+        fromContentsOf source: Slice<UnsafeMutableBufferPointer<T>>
+    ) {
+        moveAppend(fromContentsOf: UnsafeMutableBufferPointer(rebasing: source))
+    }
+}
+
+extension OutputBuffer<UInt8> /* where T: BitwiseCopyable */ {
+
+    mutating func appendBytes<Value /*: BitwiseCopyable */>(
+        of value: borrowing Value, as: Value.Type
+    ) {
+        let (q,r) = MemoryLayout<Value>.stride.quotientAndRemainder(
+            dividingBy: MemoryLayout<T>.stride
+        )
+        precondition(
+            r == 0, "Stride of Value must be divisible by stride of Element"
+        )
+        precondition(
+            (capacity &- initialized) >= q,
+            "buffer cannot contain every byte of value."
+        )
+        let p = UnsafeMutableRawPointer(start.advanced(by: initialized))
+        p.storeBytes(of: value, as: Value.self)
+        initialized &+= q
+    }
+}
+
+extension OutputBuffer {
+    var initializedPrefix: /*borrowed*/ BufferView<T> {
+        /* _read */ get /* borrowing(self) */ {
+            /* yield */ return BufferView(
+                unsafeBufferPointer: .init(start: start, count: initialized)
+            ).unsafelyUnwrapped
+        }
+    }
+
+    func withBufferView<R>(_ body: (BufferView<T>) throws -> R) rethrows -> R {
+        let view = BufferView<T>(
+            unsafeBufferPointer: .init(start: start, count: initialized)
+        ).unsafelyUnwrapped
+        return try body(view)
+    }
+}
+
+extension OutputBuffer {
+
+    consuming func relinquishBorrowedMemory() -> UnsafeMutableBufferPointer<T> {
+        let start = self.start
+        let initialized = self.initialized
+        discard self
+        return .init(start: start, count: initialized)
+    }
+}
+
+extension Array {
+
+    init(
+        capacity: Int,
+        initializingWith initializer: (inout OutputBuffer<Element>) throws -> Void
+    ) rethrows {
+        try self.init(
+            unsafeUninitializedCapacity: capacity,
+            initializingWith: { (buffer, count) in
+                var output = OutputBuffer(
+                    initializing: buffer.baseAddress.unsafelyUnwrapped,
+                    capacity: buffer.count
+                )
+                try initializer(&output)
+                let initialized = output.relinquishBorrowedMemory()
+                assert(initialized.baseAddress == buffer.baseAddress)
+                count = initialized.count
+            }
+        )
+    }
+}
+
+extension String {
```

Comment:
An organizational thing, but I think I would prefer the extensions for String and Data go with files that are closer those types than this one. Otherwise this file tends to become a 'dumping ground' for extensions.

Reply:
I agree with that, though littering `String` and `Data`'s files with experimental extensions seems undesirable as well.